File: store-service/internal/pricing/pricing_test.go
package pricing

import (
	"math"
	"slices"
	"testing"

	"duckstore/store-service/internal/packaging"
)

const eps = 0.001

func almostEqual(a, b float64) bool {
	return math.Abs(a-b) < eps
}

func TestCalculate(t *testing.T) {
	cases := []struct {
		name                string
		req                 Request
		expectedTotal       float64
		expectedDetailNames []string
	}{
		{
			// Covers: base, cardboard adjustment (-1%), other-country tax (+15%),
			// land shipping (+$10/unit). No volume discount (qty <= 100).
			//  50 * 0.99 * 1.15 + 100 = 156.925
			name: "small_cardboard_other_land",
			req: Request{
				Quantity:     10,
				UnitPrice:    5,
				Material:     packaging.Cardboard,
				Country:      "Germany",
				ShippingMode: packaging.Land,
			},
			expectedTotal:       156.925,
			expectedDetailNames: []string{"base", "material:cardboard", "country:other", "shipping:land"},
		},
		{
			// Covers: volume discount trigger (qty 101 > 100).
			// 101 * 0.8 * 0.99 * 1.15 + 1010 = 1101.9908
			name: "volume_discount_101_units",
			req: Request{
				Quantity:     101,
				UnitPrice:    1,
				Material:     packaging.Cardboard,
				Country:      "Germany",
				ShippingMode: packaging.Land,
			},
			expectedTotal:       1101.9908,
			expectedDetailNames: []string{"base", "volume_discount", "material:cardboard", "country:other", "shipping:land"},
		},
		{
			// Covers: wood (+5%), USA (+18%), sea (+$400 flat).
			// 2000 * 0.8 * 1.05 * 1.18 + 400 = 2382.4
			name: "wood_usa_sea",
			req: Request{
				Quantity:     200,
				UnitPrice:    10,
				Material:     packaging.Wood,
				Country:      "USA",
				ShippingMode: packaging.Sea,
			},
			expectedTotal:       2382.4,
			expectedDetailNames: []string{"base", "volume_discount", "material:wood", "country:usa", "shipping:sea"},
		},
		{
			// Covers: plastic (+10%), Bolivia (+13%), air normal (qty <= 1000).
			// 500 * 1.10 * 1.13 + 30*50 = 2121.5
			name: "plastic_bolivia_air_normal",
			req: Request{
				Quantity:     50,
				UnitPrice:    10,
				Material:     packaging.Plastic,
				Country:      "Bolivia",
				ShippingMode: packaging.Air,
			},
			expectedTotal:       2121.5,
			expectedDetailNames: []string{"base", "material:plastic", "country:bolivia", "shipping:air"},
		},
		{
			// Covers: India (+19%), air bulk (qty > 1000, -15% on shipping).
			// 1500 * 0.8 * 1.10 * 1.19 + 30*1500*0.85 = 1570.80 + 38250 = 39820.80
			name: "india_air_bulk_over_1000",
			req: Request{
				Quantity:     1500,
				UnitPrice:    1,
				Material:     packaging.Plastic,
				Country:      "India",
				ShippingMode: packaging.Air,
			},
			expectedTotal:       39820.80,
			expectedDetailNames: []string{"base", "volume_discount", "material:plastic", "country:india", "shipping:air"},
		},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			result := Calculate(c.req)
			if !almostEqual(result.Total, c.expectedTotal) {
				t.Errorf("Total = %.4f, want %.4f", result.Total, c.expectedTotal)
			}
			names := make([]string, len(result.Details))
			for i, d := range result.Details {
				names[i] = d.Name
			}
			if !slices.Equal(names, c.expectedDetailNames) {
				t.Errorf("detail names = %v, want %v", names, c.expectedDetailNames)
			}
		})
	}
}

File: backend/src/pricing/pricing.test.js
import { describe, it, expect } from "vitest";
import { calculate } from "./pricing.js";
import { PricingService } from "./service.js";
import { MATERIALS, SHIPPING_MODES } from "../packaging/packaging.js";

const EPS = 0.01;
const almost = (a, b) => Math.abs(a - b) < EPS;

function assertTotal(result, expected) {
  if (!almost(result.total, expected)) {
    throw new Error(`total ${result.total} !≈ ${expected}`);
  }
}

describe("calculate — happy paths", () => {
  it("base * quantity, wood + USA + air (the canonical 211.95 case)", () => {
    // 5 @ 10 = 50 → wood +5% = 52.5 → USA +18% = 61.95 → air 30×5 = 150 → 211.95
    const r = calculate({
      quantity: 5,
      unitPrice: 10,
      material: MATERIALS.Wood,
      country: "USA",
      shippingMode: SHIPPING_MODES.Air,
    });
    assertTotal(r, 211.95);
    expect(r.details.map((d) => d.name)).toEqual([
      "base",
      "material:wood",
      "country:usa",
      "shipping:air",
    ]);
  });

  it("volume discount triggers above threshold", () => {
    // 200 @ 1 = 200 → -20% = 160 → wood +5% = 168 → USA +18% = 198.24 → air 30×200 = 6000 → 6198.24
    const r = calculate({
      quantity: 200,
      unitPrice: 1,
      material: MATERIALS.Wood,
      country: "USA",
      shippingMode: SHIPPING_MODES.Air,
    });
    const names = r.details.map((d) => d.name);
    expect(names).toContain("volume_discount");
    assertTotal(r, 6198.24);
  });

  it("air bulk discount kicks in above 1000 units", () => {
    // quantity=1001 → air = 30×1001 × (1 - 0.15) = 25525.5
    const r = calculate({
      quantity: 1001,
      unitPrice: 0.01,           // keep pre-shipping small so air is the visible chunk
      material: MATERIALS.Wood,  // wood +5%
      country: "USA",
      shippingMode: SHIPPING_MODES.Air,
    });
    const air = r.details.find((d) => d.name === "shipping:air");
    expect(almost(air.amount, 25525.5)).toBe(true);
  });

  it("sea is flat $400 regardless of quantity", () => {
    const r = calculate({
      quantity: 10,
      unitPrice: 10,
      material: MATERIALS.Wood,
      country: "USA",
      shippingMode: SHIPPING_MODES.Sea,
    });
    expect(r.details.find((d) => d.name === "shipping:sea").amount).toBe(400);
  });

  it("land = 10 per unit", () => {
    const r = calculate({
      quantity: 7,
      unitPrice: 10,
      material: MATERIALS.Plastic,
      country: "Bolivia",
      shippingMode: SHIPPING_MODES.Land,
    });
    expect(r.details.find((d) => d.name === "shipping:land").amount).toBe(70);
  });
});

describe("calculate — country tax variants", () => {
  it.each([
    ["USA",     0.18, "country:usa"],
    ["Bolivia", 0.13, "country:bolivia"],
    ["India",   0.19, "country:india"],
    ["France",  0.15, "country:other"],    // default
    ["",        0.15, "country:other"],
  ])("%s → +%s (%s)", (country, rate, name) => {
    const r = calculate({
      quantity: 1,
      unitPrice: 100,
      material: MATERIALS.Cardboard,   // -1% so the country line is visible
      country,
      shippingMode: SHIPPING_MODES.Land,
    });
    const tax = r.details.find((d) => d.name === name);
    expect(tax).toBeDefined();
    // pre-tax total is 99 (100 - 1% cardboard), tax = 99 * rate
    expect(almost(tax.amount, 99 * rate)).toBe(true);
  });
});

describe("PricingService", () => {
  it("exposes calculate and delegates to the factory", () => {
    const svc = new PricingService();
    const r = svc.calculate({
      quantity: 5,
      unitPrice: 10,
      material: MATERIALS.Wood,
      country: "USA",
      shippingMode: SHIPPING_MODES.Air,
    });
    assertTotal(r, 211.95);
    expect(svc.entityName).toBe("pricing");
  });
});
